### ISSUE TITLE
Example code to read an in-memory JSON-LD document?

### ISSUE THREAD

**Issue by colin-young** [LDflex/LDflex#81]
I'm building a POC and don't have my data published anywhere (plus corporate environment so lots of network security issues anyway), so I'd like to be able to read from an in-memory JSON-LD string/object. I think I've managed to get it read in through `JsonLdParser`, created a `ComunicaEngine` passing in the parser as the data and then set up a path from `PathFactory` passing in the `@context` and the queryEngine. A simplified version of the data looks like:

```
{
  "@context": "http://schema.org/",
  "@graph": [{
    "@id": "https://example.org/Person/JaneDoe",
    "@type": "Person",
    "name": "Jane Doe",
    "jobTitle": "Professor",
    "telephone": "(425) 123-4567",
    "url": "http://www.janedoe.com"
  },{
    "@id": "https://example.org/Person/JohnDoe",
    "@type": "Person",
    "name": "John Doe",
    "jobTitle": "Professor",
    "telephone": "(425) 234-5678",
    "url": "http://www.johndoe.com"
  }]
}
```

And now I'm stuck. Assuming the previous set things up correctly, I don't know how to get at an item in my data that I can then pass in to src using the solid/react-components library. e.g. I'd like to display a list of the names. Can anyone provide some tips or point me to a complete working example?

**Comment by jeswr** [comment]
I don't think the current ldflex-comunica config is really setup to do this (see https://github.com/LDflex/LDflex-Comunica/issues/3). 

The workaround that I have been using for in memory data is to create a store using the n3 package and then override the engine
```ts
import { newEngine } from '@comunica/actor-init-sparql-rdfjs';
import ComunicaEngine from '@ldflex/ldflex-comunica';
import { Store } from 'n3';

const store = new Store();
const quads = /* Parsed quads using something like https://github.com/rdfjs-base/parser-jsonld */
store.addQuads(quads);

const queryEngine = new ComunicaEngine();
queryEngine._engine = newEngine();
queryEngine._sources = [store]
```

otherwise, if you want to read from the file directly you could use

```ts
import { newEngine } from '@comunica/actor-init-sparql-file';
import ComunicaEngine from '@ldflex/ldflex-comunica';

const queryEngine = new ComunicaEngine();
queryEngine._engine = newEngine();
queryEngine._sources = [/* path to file */]
```

I did create a *very hacky* workaround [here](https://github.com/jeswr/LDflex-Comunica/blob/auto-local-query/src/ComunicaEngine.ts); which automatically deteremines which of those engines to apply based on the input source type (I certainly would not use it for production).

**Comment by colin-young** [comment]
Thanks, that is very helpful. I think I can just use https://github.com/rubensworks/rdf-store-stream.js to get a store from the https://github.com/rubensworks/jsonld-streaming-parser.js library. Now my issue appears to be figuring out how to set the JSON-LD content for the parser from code (i.e. not reading a local file).

**Comment by jeswr** [comment]
Can't you just import/require it since it as JSON so it becomes part of your static browser files (I'm on my phone currently so can't test this)?

```ts
import data from 'my-data.jsonld'

Parser.parse(JSON.stringify(data))
```

Otherwise you could just export the data as a constant string from a .js/.ts file and parse that string.


```ts
export ​default `{
  "@context": "http://schema.org/",
  "@graph": [{
    "@id": "https://example.org/Person/JaneDoe",
    "@type": "Person",
    "name": "Jane Doe",
    "jobTitle": "Professor",
    "telephone": "(425) 123-4567",
    "url": "http://www.janedoe.com"
  },{
    "@id": "https://example.org/Person/JohnDoe",
    "@type": "Person",
    "name": "John Doe",
    "jobTitle": "Professor",
    "telephone": "(425) 234-5678",
    "url": "http://www.johndoe.com"
  }]
}`
```

PS not sure of you're use case but you can always use electron if you need your app to be able to work with the local fs.

**Comment by jeswr** [comment]
For `path.create` you are providing whatever subject node you want to focus on. For instance the following will log all objects matching the pattern `<http://example.org/John> <http://example.org/friend> ?object`

```ts
const path = new PathFactory(...)
const John = path.create({ subject : namedNode('http://example.org/John') })

(async () => {
  for await (const friend of John['http://example org/friend'])
    console.log(friend)
})

```

There are some cases where the input doesn't matter as discussed here.
https://github.com/LDflex/LDflex/issues/75#issuecomment-742818362

**Comment by colin-young** [comment]
I switched from the streaming json-ld parser to the one you suggested in your first response, and now I'm at least getting somewhere. According to the [playground](https://json-ld.org/playground/#startTab=tab-nquads&json-ld=%7B%22%40context%22%3A%22https%3A%2F%2Fschema.org%2F%22%2C%22%40graph%22%3A%5B%7B%22%40id%22%3A%22https%3A%2F%2Fexample.org%2FPerson%2FJaneDoe%22%2C%22%40type%22%3A%22Person%22%2C%22name%22%3A%22Jane%20Doe%22%2C%22jobTitle%22%3A%22Professor%22%2C%22telephone%22%3A%22(425)%20123-4567%22%2C%22url%22%3A%22http%3A%2F%2Fwww.janedoe.com%22%7D%2C%7B%22%40id%22%3A%22https%3A%2F%2Fexample.org%2FPerson%2FJohnDoe%22%2C%22%40type%22%3A%22Person%22%2C%22name%22%3A%22John%20Doe%22%2C%22jobTitle%22%3A%22Professor%22%2C%22telephone%22%3A%22(425)%20234-5678%22%2C%22url%22%3A%22http%3A%2F%2Fwww.johndoe.com%22%7D%5D%7D&frame=%7B%7D&context=%7B%22%40context%22%3A%22https%3A%2F%2Fschema.org%2F%22%7D), I have `<https://example.org/Person/JohnDoe> <http://schema.org/name> "John Doe" .`, but attempting:

```
const parserJsonld = new ParserJsonld();
const input = new Readable({
  read: () => {
    input.push(jsonLdString);
    input.push(null);
  }
});

const output = parserJsonld.import(input);
const store = new Store();
store.addQuads(output);

const queryEngine = new ComunicaEngine();
queryEngine._engine = newEngine();
queryEngine._sources = [store];

const path = new PathFactory({
  context: { "@context": "https://schema.org/" },
  queryEngine,
});

const participantList = path.create({
  subject: namedNode("https://example.org/Person/JohnDoe")
});

(async () => {
  for await (const person of participantList)
    console.log(`name: ${person["http://schema.org/name"]}`);
})();
```

just displays `undefined` (twice).

My mental model of what should be going on, based on @jeswr 's latest comment (I apologize in advance for mangling terminology - I'm just starting to wrap my head around this stuff) is:
- from my path I get an "object" which has several "properties", or, rather, I get the collection of N-quads that are identified by `<https://example.org/Person/JohnDoe>`, so I expect 5 of them for my example. 
- Then I expect that the `${person["http://schema.org/name"]}` will find the quad in that collection that has the predicate `http://schema.org/name`
- and give me the `?object` which, in this case, I expect to be the literal string "John Doe".
 
But that doesn't appear to be what is happening. If I just output the `person` in my log statement what I get is the value of the `?subject` (`@id` in JSON-LD, or literally `https://example.org/Person/JohnDoe`). Which doesn't seem correct. Or terribly useful. In fact, it seems as though it just returns whatever it is that I set the `subject` to. I can enter any old garbage for the `subject` (including nonsense that isn't a valid URI or even a string that occurs in my data) and it will output exactly what I entered. Twice (even after modifying the input to have 3 `person` definitions).

**Comment by jeswr** [comment]
The issue with the current code is in your last few lines. The `for await` structure is used to iterate through if you already have some sort of path/predicate applied (like with friends above). Here; participantList is just the ldflex object for John Doe so to get just John Does name you can do this:

```ts
​const​ ​person​ ​=​ ​path​.​create​(​{​
  ​subject​: ​namedNode​(​"https://example.org/Person/JohnDoe"​)​
​}​)​;​

​(​async​ ​(​)​ ​=>​ ​{​
    ​console​.​log​(​`name: ​${​await person​[​"http://schema.org/name"​]​}​`​)​;​
​}​)​(​)​;
```

Note that `await` is needed to resolve his name as well.

On the other have; if you wanted to get the name of every subject in the graph you would go

```ts
​const​ ​participantList​ ​=​ ​path​.​create​(​{​
  ​subject​: ​namedNode​(​"https://example.org/Person/JohnDoe"​)​
​}​)​;​

​(​async​ ​(​)​ ​=>​ ​{
  for await (const person of participantList.subjects)
    ​console​.​log​(​`name: ​${​await person​[​"http://schema.org/name"​]​}​`​)​;​
​}​)​(​)​;
```

**Comment by jeswr** [comment]
This Cheatsheet given in the readme goes through most of this.

**Comment by colin-young** [comment]
I've seen the cheatsheet (when I originally looked at it I quickly realized I wasn't at the stage where it would be useful, and then didn't realize that I had now reached that stage), but even with the corrections you've suggested, I'm still seeing undefined.

I added the following to try and get some insight into the parsing:

```
output.on("data", (quad) => {
    console.log(
        `${quad.subject.value} - ${quad.predicate.value} - ${quad.object.value}`
    );
});
```

but nothing is being output to the console. Since I also appear to be getting whatever I supply as the subject, it's starting to look as though something has gone wrong with the parsing of the json-ld string, which I've more or less confirmed by setting it to an empty string and getting the same results. Investigating the loading/parsing through ParserJsonLd now.

**Comment by jeswr** [comment]
I'm not familiar with the API of that specific parser but I'll quickly tinker with it later today too see if I can debug that for you.

My guess is`.import` is not being used correctly.

**Comment by colin-young** [comment]
I'm thinking it might be getting my json string into a stream. I'm following the example at the very bottom of https://github.com/rdfjs-base/parser-jsonld#readme, but I don't ever see any quads logged in the console, nor do I hit any breakpoints inside the code for creating the `Readable`. I have attempted with both jsonld parsing libraries with identical results (i.e. https://github.com/rubensworks/jsonld-streaming-parser.js and https://github.com/rdfjs-base/parser-jsonld).

I appreciate all the assistance.

**Comment by jeswr** [comment]
I've created a minimal working example here https://github.com/jeswr/playground. The key bug with the original version was that `store.addQuads` accepts an array of quads but you were inputting a stream of quads instead.

In the example repo there is a test.ts file (I prefer to work in typescript as it catches errors like this where the input type doesn't match the parameter type). But I have also transpiled the result to a `test.js` file if you prefer that (you can just ignore the transpiled code before the import/require statements).

**Comment by colin-young** [comment]
Thanks for the excellent example!

It turns out there were 2 main issues with my code. The first was me (as you identified -- I only discovered streams in js the other day so I am very much not up to speed with them). The second is my locked-down, proxied, corporate environment. I was able to clone and run the example in a matter of minutes on my personal machine. On my work machine it's a self-signed certificate error, followed by, after fixing, ECONNRESET when attempting to fetch the context from schema.org. Oddly, there are no errors shown when implementing the code in a jsx file, but they are shown in the provided example command line.

I have vague memories of running into that issue about a year ago when I was experimenting with jena/fuseki and I think it has to do with our corporate security. The odd thing is at some point in this experiment I did see quads being output to the console, meaning it was parsing things at some point. I think I remember seeing an issue about the JSON-LD parser not supporting local context files (to allow it to run disconnected from the network), but I don't think that was implemented yet, and I can't seem to find it now.

In any case, this question has been thoroughly answered, so I will close it. Thanks again!
